fgpyo/sam/__init__.py

```diff
@@ -1231,3 +1250,285 @@ class SamOrder(enum.Enum):
+            ValueError: If `tag` is `XA` and `value` does not have 6 comma-separated fields.
+            ValueError: If `tag` is `SA` and `value` does not have '+' or '-' as a strand.
+            ValueError: If `tag` is `XA` or `XB` and position is not a stranded integer.
+        """
+        if ";" in value:
+            raise ValueError(f"Cannot parse a multi-value string! Found: {value} for tag {tag}")
+
+        fields: list[str] = value.rstrip(",").split(",")
+
+        if tag == "SA" and len(fields) == 6:
+            reference_name, reference_start, strand, cigar, mapq, edit_distance = fields
+
+            if strand not in ("+", "-"):
+                raise ValueError(f"The strand field is not either '+' or '-': {strand}")
+
+            return cls(
+                reference_name=reference_name,
+                reference_start=int(reference_start) - 1,
+                is_forward=strand == "+",
+                cigar=Cigar.from_cigarstring(cigar),
+                mapping_quality=None if mapq is None else int(mapq),
+                edit_distance=int(edit_distance),
+                is_secondary=False,
+                is_supplementary=True,
+            )
+
+        elif tag in ("XA", "XB") and (num_fields := len(fields)) in (4, 6):
+            if num_fields == 4:
+                mapq = None
+                alignment_score = None
+                reference_name, stranded_start, cigar, edit_distance = fields
+            else:
+                reference_name, stranded_start, cigar, edit_distance, alignment_score, mapq = fields
+
+            if len(stranded_start) <= 1 or stranded_start[0] not in ("+", "-"):
+                raise ValueError(f"The stranded start field is malformed: {stranded_start}")
+
+            return cls(
+                reference_name=reference_name,
+                reference_start=int(stranded_start[1:]) - 1,
+                is_forward=stranded_start[0] == "+",
+                cigar=Cigar.from_cigarstring(cigar),
+                mapping_quality=None if mapq is None else int(mapq),
+                edit_distance=int(edit_distance),
+                alignment_score=None if alignment_score is None else int(alignment_score),
+                is_secondary=True,
+                is_supplementary=False,
+            )
+
+        else:
+            raise ValueError(f"{tag} tag value has the incorrect number of fields: {value}")
+
+    @classmethod
+    def many_from_primary(cls, primary: AlignedSegment) -> list[Self]:
+        """Build all auxiliary alignments for a given primary alignment.
+
+        Args:
+            primary: The primary alignment to build auxiliary alignments from.
+
+        Raises:
+            ValueError: If the input record is a secondary or supplementary alignment.
+        """
+        if primary.is_secondary or primary.is_supplementary:
+            raise ValueError("Cannot build auxiliary alignments from a non-primary alignment!")
```

Comment:
But you can!

Reply:
Just to add to this, after Nils' comments in slack yesterday, I've found it helpful to disambiguate the use of "primary" with the following distinctions:

- **"primary"** refers to _"non-secondary"_, meaning the best (or arbitrarily selected) among multiple alignments .
- **"representative"** refers to _"non-supplementary"_, meaning the alignment record chosen to represent a chimeric alignment

Reply:
I can't speak to how secondary alignments are typically represented, but I believe that supplementary alignments do typically include the `SA` tag with aux-representations of the representative and any other supplementary alignments